### ISSUE TITLE
correct types for pattern matchers

### ISSUE THREAD

**Issue by turadg** [Agoric/Agoric#6206]
## What is the Problem Being Solved?

While working on this again, documenting `.optional` would help (https://github.com/Agoric/agoric-sdk/pull/6202/files#r971342637 )

I wondered why code could use it without the type and it's because `M.call()` returns https://github.com/Agoric/agoric-sdk/blob/870bd222933a57559b9fc9abd38b4e2c1e9ac69e/packages/store/src/types.js#L611-L612

Relevant
- https://github.com/Agoric/agoric-sdk/issues/6160 

## Description of the Design

No `any` for pattern matcher types.

**Comment by erights** [comment]
FWIW, this is implemented at https://github.com/Agoric/agoric-sdk/blob/9b786284b4ff759dd050c701829df5fd0cf41749/packages/store/src/patterns/patternMatchers.js#L1537-L1543 and used several times in https://github.com/Agoric/agoric-sdk/blob/master/packages/ERTP/src/typeGuards.js

I'm actually surprised I didn't write any comments explaining this at all. I certainly agree it is needed.

Also `rest`

**Comment by erights** [comment]
@turadg  How much of this issue is addressed by https://github.com/endojs/endo/pull/1715 ?

Note that https://github.com/endojs/endo/pull/1715 is merged into endo. But an endo with that change is not yet released, and therefore not yet available to agoric-sdk.

**Comment by turadg** [comment]
When I link HEAD of `@endo/patterns`, it appears to be solved by 1715. 
<img width="603" alt="Screenshot 2023-08-28 at 4 34 18 PM" src="https://github.com/Agoric/agoric-sdk/assets/21505/6bf88b83-284d-4bec-87d2-100621fc3b69">

Unfortunately the type when importing `M` from `@agoric/store` has gotten worse, in that the whole thing is `any`. I don't know why but I expect https://github.com/Agoric/agoric-sdk/pull/7949 would fix it

**Comment by erights** [comment]
Oh sorry! I lost track of #7949. Looking now.

Are you waiting on me for anything else?

**Comment by turadg** [comment]
No worries! Wasn't waiting on 7949 since it was low prio. If it was high I would have pestered you 😉

**Comment by erights** [comment]
Good! In any case, LGTMed that one. Please let me know whether it fixes the issue here.
